### PR TITLE
[VIRTS-1998] Config API v2

### PR DESCRIPTION
**Pull request by bworrell**
## Description
Added an API for configuration and other utilities/middleware that can be used for other API endpoints as we continue to build out API v2.

Notes:
* Sensitive configuration items (users, keys, passwords, etc.) are filtered out of responses and/or rejected in requests. This is currently controlled in code via a global variable--we may want to break this out into some sort of configuration later, or make it possible to annotate certain config items as sensitive/api-excluded.
* Responses are just generic JSON documents w/ variable keys (we might want to change this) so there is no defined response schema. Because there is no response schema (marshmallow schema), the response is not included in the generated swagger docs.

## New endpoints:
* `GET /api/v2/config/:name`
  * Returns the entire configuration matching `name` (e.g., `main` or `agents`).
* `PATCH /api/v2/config/main` 
  * Update the `main` configuration
* `PATCH /api/v2/config/agents`
  * Update the `agents` configuration   

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
* Added unit tests for config api manager
* Added unit tests for json validation middleware
* Confirmed endpoints appear in swagger documentation (see note in description about the omission of a response schema)
* Confirmed that endpoints return 401 if not authenticated
* Hit the endpoints in Postman w/ api key auth
  * Tested successful main config update
  * Tested successful agent config update
  * Tested successful GET of main config
  * Tested successful GET of agents config
  * Tested trying to GET an invalid config (got a 404 json response)
  * Tested trying to update 'sensitive' field (got 403 response)
  * Tested trying to make an update to an unknown field (got 400 response)
  * Tested trying to make an update w/ wrong data type (got 400 response)

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: app/api/v2/managers/config_api_manager.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                prop=prop,
+                value=value
+            )
+
+    async def update_global_agent_config(self, sleep_min=None, sleep_max=None, watchdog=None, untrusted_timer=None,
+                                         implant_name=None, bootstrap_abilities=None, deadman_abilities=None):
+
+        set_config = self._config_interface.set_config
+
+        if sleep_min is not None:
```
Comment (bworrell) [inline app/api/v2/managers/config_api_manager.py:95]: I don't like that we don't perform any validation on these values (aside from the datatype validation performed by marshmallow). I am gonna add some validation--just need to decide where the best place is do do that

**Review by b17zr — CHANGES_REQUESTED** [review]
Looks really good! I only have nits on quotes.

File: tests/api/v2/test_validation.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def test_check_not_empty_string_throws_on_non_string():
+    with pytest.raises(errors.DataValidationError):
+        validation.check_not_empty_string(None)
+
+
+def test_check_not_empty_string_throws_on_empty_string():
+    with pytest.raises(errors.DataValidationError):
+        validation.check_not_empty_string("")
```
[the remark this answers is not in the document]
Comment (bworrell) [reply tests/api/v2/test_validation.py:44]: I'll update that!

File: tests/api/v2/managers/test_config_api_manager.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+    with pytest.raises(errors.DataValidationError):
+        await manager.update_global_agent_config(untrusted_timer=-1)
+
+
+async def test_update_global_agent_config_throws_validation_error_bad_implant_name(base_world, data_svc):
+    manager = ConfigApiManager(data_svc)
+
+    with pytest.raises(errors.DataValidationError):
+        await manager.update_global_agent_config(implant_name="")
```
[the remark this answers is not in the document]
Comment (bworrell) [reply tests/api/v2/managers/test_config_api_manager.py:204]: image

[the remark this answers is not in the document]
Comment (bworrell) [reply tests/api/v2/managers/test_config_api_manager.py:204]: i'll fix it :D

**Review by b17zr — DISMISSED** [review]
LGTM!

File: app/api/v2/managers/config_api_manager.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+SENSITIVE_CONFIG_PROPS = frozenset([
+    'ability_refresh',
+    'api_key_blue',
+    'api_key_red',
+    'app.contact.gist',
+    'app.contact.tunnel.ssh.host_key_file',
+    'app.contact.tunnel.ssh.host_key_passphrase',
+    'app.contact.tunnel.ssh.user_name',
+    'app.contact.tunnel.ssh.user_password',
```
Comment (wbooth) [inline app/api/v2/managers/config_api_manager.py:16]: drop app.contact

File: app/api/v2/managers/config_api_manager.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
+import logging
+from typing import List
+
+from app.api.v2 import validation
+from app.utility.base_world import BaseWorld
+
+
+SENSITIVE_CONFIG_PROPS = frozenset([
+    'ability_refresh',
```
Comment (wbooth) [inline app/api/v2/managers/config_api_manager.py:9]: drop ability_refresh?

**Review by b17zr — APPROVED** [review]

**Review by wbooth — APPROVED** [review]

**Pull request merged**